provide compat_platform replacement for deprecated platform.dist() Fixes #30489

alikins opened the pull request:
##### SUMMARY
Add compat_platform.py to provide deprecated platform.dist
    
The 'dist' and 'linux_distribution' methods in the python stdlib module 'platform.py' are deprecated.
    
For compatibility, this adds a copy of a version of platform.py to module_utils in lib/ansible/module_utils/compat_platform.py
    
Code that was using platform.dist() or platform.linux_distribution() has been updated to use ansible.module_utils.compat_platform instead.
    
Other uses of platform.py are left as is.
    
Fixes #30489

<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request


##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
lib/ansible/module_utils/facts/system/distribution.py
lib/ansible/module_utils/basic.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.0 (compat_platform_30489 f6b9a84be2) last updated 2018/01/24 15:14:53 (GMT -400)
  config file = /home/adrian/src/ansible/ansible.cfg
  configured module search path = [u'/home/adrian/ansible/my-modules']
  ansible python module location = /home/adrian/src/ansible/lib/ansible
  executable location = /home/adrian/src/ansible/bin/ansible
  python version = 2.7.14 (default, Jan 17 2018, 14:28:32) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]


```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

alikins committed:
Add simple intg tests for compat_platform py module

alikins committed:
Add compat_platform.py to provide deprecated platform.dist

The 'dist' and 'linux_distribution' methods in the python
stdlib module 'platform.py' are deprecated.

For compatibility, this adds a copy of a version of platform.py
to module_utils in lib/ansible/module_utils/compat_platform.py

Code that was using platform.dist() or platform.linux_distribution()
has been updated to use ansible.module_utils.compat_platform instead.

Other uses of platform.py are left as is.

Fixes #30489

alikins commented:
Should mention that the version of compat_platform.py is based off the one from python 3.6.4 but
with quick backporting to be python2 compatible. 

Some of the changes should be harmless. Some were just removal of code we shouldn't need (mostly the 'run regexes on sys.executable binary to detect libc info' bits) mostly because of some py3 only regex options but could likely be updated to work for both.  

There is at least one FIXME at the moment related to file encoding that should get cleaned up, possibly using _text.* methods.

There is still a lot of code in there that could/should be removed since it is now redundant with non-deprecated parts of platform.py, but wanted to start with a version as close to upstream as possible.

alikins commented:
Looking at the CI failures, it looks like Ubuntu may use a patched version of platform.py

```
2018-01-24 20:49:22 diff -u of builtin and vendored
2018-01-24 20:49:22 --- /tmp/tmp.EZxQzu1mvu/builtin_output.txt	2018-01-24 20:49:22.797424404 +0000
2018-01-24 20:49:22 +++ /tmp/tmp.EZxQzu1mvu/vendored_output.txt	2018-01-24 20:49:22.841422211 +0000
2018-01-24 20:49:22 @@ -1 +1 @@
2018-01-24 20:49:22 -Linux-3.19.0-51-generic-x86_64-with-Ubuntu-14.04-trusty
2018-01-24 20:49:22 +Linux-3.19.0-51-generic-x86_64-with-debian-jessie-sid
```

I don't see any other test failures for other platforms (yet) but I kind of expect other distros will have patches platform.py as well. But that was what the intg test was designed to catch, so I guess it worked.

alikins commented:
Looking at one version of ubuntu patches to platform.py[1]:

http://bazaar.launchpad.net/~doko/python/pkg2.7-debian/view/head:/patches/platform-lsbrelease.diff

Does look Ubuntu patches platform.py so I guess we need those changes as well.

[1] I know nothing about ubuntu packaging, what is the best place to look for 'here are the patches ubuntu applies to UpstreamProject for ubuntu 16.04 for ex' ?

alikins committed:
remove shebang from module

alikins committed:
Add compat_platform.py to sanity ignore lists for now

alikins committed:
Apply Ubuntu's patch to platform.py to compat_platform.py

Changes are from the Ubuntu python packages patch to
platform.py to recognize Ubuntu. In particular this patch:

http://bazaar.launchpad.net/~doko/python/pkg2.7-debian/view/head:/patches/platform-lsbrelease.diff

alikins committed:
cleanup and deverbose-ify compat_platform/runme.sh

alikins committed:
add compat_platform to pep8 skip.txt

rm from legacy-files.txt?

alikins committed:
make ubuntu regexes r'strings' as suggested by pylint

mattclay commented on test/sanity/pylint/ignore.txt:
```diff
@@ -1,3 +1,5 @@
+lib/ansible/module_utils/compat_platform.py multiple-imports
+lib/ansible/module_utils/compat_platform.py ansible-format-automatic-specification
```
Why are ignore/skip entries being added?

mattclay replied:
In particular the `ansible-format-automatic-specification` ignore should not be used, as that rule indicates a Python 2.6 compatibility issue.

alikins replied:
The file started as a copy of 'platform.py' from python3.6  so I'm trying to keep it as close to upstream as possible for now. It already has a few tweaks to get it py2/py3 compatible though.

Sounds like I need to update the code that triggers ansible-format-automatic-specification errors as well to get it to be py2.6 compatible.

alikins replied:
I also need to add some unit tests to exercise it, since it passed 2.6 unit tests. I'll take a look a the coverage and see what it needs. Most of the module is unused at the moment, so it's possible tests covered some of it but not the py26 incompat parts though I expect it is just not covered in tests.

sivel replied:
The area of the code that failed that check, only happens on windows, so it might not be super easy to cover.

```diff
 +    # getwindowsversion() reflect the compatibility mode Python is
 +    # running under, and so the service pack value is only going to be
 +    # valid if the versions match.
 +    if winver[:2] == (maj, min):
 +        try:
 +            csd = 'SP{}'.format(winver.service_pack_major)
```

mattclay replied:
A unit test with some mocking should be able to cover that, but probably not "super easy".

alikins replied:
@sivel And it looks like that code is only hit if using the compat_platform.win32_ver() which no code should be doing. Any code that needs that should continue to use the regular python platform.py since only the 'dist' and 'linux_distribution' methods are deprecated.

I can just remove all of the unused code here. 

I was thinking it might be easier to keep in sync with upstream pythons platform.py if we kept compat_platform.py as close as possible to upstream.

But now I think I'll just cut it down to just the code needed for the dist related functions.
I don't think keeping it exactly like upstream will buy us much. It's really not that much code in there that we need, and the upstream platform.py doesn't change very much (and the we need will be removed upstream after 3.7). And the code has already needed some py2/py3 compat tweaks, plus the ubuntu specific changes that isnt upstream.

ie, I'm going to remove all of the windows related code that fails sanity and anything else thats not needed.

alikins committed:
add test_compat_platform.py unit tests

mostly stubs currently with very few asserts

alikins committed:
Remove every not related to linux_distribution()/dist()

Only dist() and linux_distribution() are deprecated, so
all other methods still exist in upstream python platform.py

alikins committed:
remove unneeded warnings and shorten comments

alikins committed:
fix default args

alikins committed:
cleaning up shadowed builtins, misc pep8 etc

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/52675/3/tests

CI failure in multiple integration tests as well, although the ubuntu1404/1 failure is unrelated.

bcoca commented:
since some distros 'fix/patch' these i would make it a fallback, use the ones we ship if we fail to use the system ones.

Commit:
compat_platform.py now passes pylint sanity

Commit:
test_compat_platform.py cleanup

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/53713/3/tests
Integration tests too: https://app.shippable.com/github/ansible/ansible/runs/53713/41/console

alikins committed:
Better intg test for compat_platform

alikins committed:
rm unit tests for null cases

not typically called that way, and the ubuntu support that
is in compat_platform and patches in some 'platform' modules
causes it to not fail as expected.

Add a unit tests that compares platform.* and compat_platform.*
where we have both

alikins committed:
rm missed empty arg test case

alikins committed:
add compat_platform to no-underscore exceptions;

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/55792/7/tests

Integration test failure is unrelated.

marcosps commented:
Would be good to have this patch merged, so it could be possible to use the Ansible's hostname module in openSUSE, since right now platform.linux_distribution() returns empty in Leap 15.0.

abadger commented:
We're going to merge a solution for 2.8 but it will be based on
https://github.com/ansible/ansible/pull/35985 rather than this one.  Going
to close this one too about confusion.

If newer opensuse is doing the wrong thing with that version, someone
should either let nir0s know so they can update their library or we can
look and see if we can put some overriding information in our code
